enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,359 @@
+which is a `kubernetes.io/tls` type secret reference. This `externalCertificate` field will enables the
+users to provide a reference to a Secret containing the serving cert/key pair that will be parsed
+and served by OpenShift router.
+
+### Workflow Description
+
+The following workflow describes the integration with third party
+certificate management solutions like cert-manager with the certificate
+reference field described under [API Extensions](#api-extensions).
+
+- The end user must have generated the serving certificate generated
+  as a pre requisite using third-party systems like cert-manager.
+- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources)
+  CR must be created in the same namespace where the Route is going to be created.
+- The end user must create a role and in the same namespace as the secret containing the certificate from earlier,
+  ```bash
+  oc create role secret-reader --verb=get,list,watch --resource=secrets --resourceName=<secret-name>
+  ```
+- The end user must create a rolebinding in the same namespace as the secret
+  and bind the router serviceaccount to the above created role.
+  ```bash
+  oc create rolebinding foo-secret-reader --role=secret-reader --serviceaccount=openshift-ingress:router --namespace=<current-namespace>
+  ```
+- To expose a user workload, the user would create a new Route with the
+  `.spec.tls.externalCertificate` referencing the generated secret that was created
+  in the previous step.
+- If the secret that is referenced exists and has a successfully generated
+  cert/key pair, the router will serve this certificate if all preconditions are met.
+
+#### Variation [optional]
+
+N.A
+
+### API Extensions
+
+A `.spec.tls.externalCertificate` field is added to Route `.spec.tls` which can be used to provide a secret name
+containing the certificate data instead of using `.spec.tls.certificate` and `spec.tls.key`.
+
+```go
+
+// TLSConfig defines config used to secure a route and provide termination
+//
+// +kubebuilder:validation:XValidation:rule="has(self.termination) && has(self.insecureEdgeTerminationPolicy) ? !((self.termination=='passthrough') && (self.insecureEdgeTerminationPolicy=='Allow')) : true", message="cannot have both spec.tls.termination: passthrough and spec.tls.insecureEdgeTerminationPolicy: Allow"
+// +openshift:validation:FeatureSetAwareXValidation:featureSet=TechPreviewNoUpgrade;CustomNoUpgrade,rule="!(has(self.certificate) && has(self.externalCertificate))", message="cannot have both spec.tls.certificate and spec.tls.externalCertificate"
+type TLSConfig struct {
+	// ...
+
+	// externalCertificate provides certificate contents as a secret reference.
+	// This should be a single serving certificate, not a certificate
+	// chain. Do not include a CA certificate. The secret referenced should
+	// be present in the same namespace as that of the Route.
+	// Forbidden when `certificate` is set.
```

Comment:
Should we mention the fact that the RBAC needs to be done by the user?

Reply:
We could do that. But I don't see a concise way to mention this detail. Either we'll only have to mention, this field requires manual rbac to be created and maybe link to docs. Putting the entire rbac (role, rolebinding, etc) would be too bloated I feel. Moreover there are couple of variations in rbac as well. 
```
- Grant router service account access to secret by secret-name (explicit rbac)
- Grant router service account access to all secrets in a fixed namespace (implicit rbac)
```

Reply:
I thought of a simple phrase without any doc references which can give a motivation to the further digging though. Something like: "Router service account needs to be granted with read-only access to this secret."

Reply:
I've added a note in `TP -> GA criteria` to update the API. Since I didn't want to start a new API update cycle. Hope this is fine.

Reply:
Yes, it's fine with me. I think we cannot have enough docs highlighting this fact. I'm sure the OCP docs will be on point but I believe that `oc explain` should be ready too.